### PR TITLE
port of OpenSSL changes to Maistra-1.2

### PR DESCRIPTION
**Pull request by dmitri-d**
* Ported OpenSSL-related changes and fixes to Maistra-1.2
* Tests are passing

### CODE REVIEW

File: source/extensions/transport_sockets/tls/context_manager_impl.h [hunk 1]
```diff
@@ -40,6 +40,7 @@ class ContextManagerImpl final : public Envoy::Ssl::ContextManager {
     return private_key_method_manager_;
   };
 
+
```
Comment (knrc) [inline source/extensions/transport_sockets/tls/context_manager_impl.h:43]: ?

File: source/extensions/transport_sockets/tls/ssl_socket.cc [hunk 2]
```diff
@@ -503,8 +504,10 @@ const std::string& SslSocketInfo::tlsVersion() const {
 }
 
 absl::optional<std::string> SslSocketInfo::x509Extension(absl::string_view extension_name) const {
+  std::cout << "AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA\n";
```
Comment (knrc) [inline source/extensions/transport_sockets/tls/ssl_socket.cc:507]: ?

File: source/extensions/transport_sockets/tls/ssl_socket.cc [hunk 3]
```diff
@@ -503,8 +504,10 @@ const std::string& SslSocketInfo::tlsVersion() const {
...
   bssl::UniquePtr<X509> cert(SSL_get_peer_certificate(ssl_.get()));
   if (!cert) {
+    std::cout << "BBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBB\n";
```
Comment (knrc) [inline source/extensions/transport_sockets/tls/ssl_socket.cc:510]: ?

File: test/common/crypto/utility_test.cc [hunk 4]
```diff
@@ -53,6 +53,7 @@ TEST(UtilityTest, TestSha256HmacWithEmptyArguments) {
   EXPECT_EQ("b613679a0814d9ec772f95d778c35fc5ff1697c493715653c6c712144292c5ad", Hex::encode(hmac));
 }
 
+// TODO (dmitri-d) re-enabled when importPublicKey OpenSSL-compatible implementation has been added
```
Comment (knrc) [inline test/common/crypto/utility_test.cc:56]: Has this been implemented?

File: test/integration/protocol_integration_test.cc [hunk 5]
```diff
@@ -420,7 +420,6 @@ TEST_P(DownstreamProtocolIntegrationTest, RetryPriority) {
   EXPECT_EQ(512U, response->body().size());
 }
 
-//
```
Comment (knrc) [inline test/integration/protocol_integration_test.cc:423]: ?

File: test/extensions/transport_sockets/tls/context_impl_test.cc [hunk 6]
```diff
@@ -1463,6 +1452,8 @@ TEST_F(ServerContextConfigImplTest, PrivateKeyMethodLoadFailureNoProvider) {
       "Failed to load incomplete certificate from ");
 }
 
+// TODO (dmitri-d) we do not support key providers under OpenSSL atm.
+/*
```
Comment (knrc) [inline test/extensions/transport_sockets/tls/context_impl_test.cc:1456]: Do we have a list of these differences?

File: source/extensions/transport_sockets/tls/ssl_socket.cc (hunk 2 above)
Comment (dmitri-d) [reply source/extensions/transport_sockets/tls/ssl_socket.cc:507]: fixed

File: source/extensions/transport_sockets/tls/ssl_socket.cc (hunk 3 above)
Comment (dmitri-d) [reply source/extensions/transport_sockets/tls/ssl_socket.cc:510]: oops, I thought I removed all of these.

File: test/common/crypto/utility_test.cc (hunk 4 above)
Comment (dmitri-d) [reply test/common/crypto/utility_test.cc:56]: Yes, removed the comments

File: test/extensions/transport_sockets/tls/context_impl_test.cc (hunk 6 above)
Comment (dmitri-d) [reply test/extensions/transport_sockets/tls/context_impl_test.cc:1456]: Not sure, there might be something in the docs (the only user-facing difference we have is lack of support for QUIC). This particular instance is an implementation detail -- the main purpose of this abstraction is to support offloading of private keys to dedicated h/w under boringssl.

Comment (dmitri-d) [reply test/extensions/transport_sockets/tls/context_impl_test.cc:1456]: Need to clean this up, though, I'll add it to tasks for 1.2 release.

File: test/extensions/transport_sockets/tls/utility_test.cc [hunk 7]
```diff
@@ -121,7 +124,7 @@ TEST(UtilityTest, GetLastCryptoError) {
 
   ERR_put_error(ERR_LIB_SSL, 0, ERR_R_MALLOC_FAILURE, __FILE__, __LINE__);
   EXPECT_EQ(Utility::getLastCryptoError().value(),
-            "error:10000041:SSL routines:OPENSSL_internal:malloc failure");
+            "error:14000041:SSL routines:SSL routines:malloc failure");
```
Comment (brian-avery) [inline test/extensions/transport_sockets/tls/utility_test.cc:127]: is this defined somewhere?

Comment (dmitri-d) [reply test/extensions/transport_sockets/tls/utility_test.cc:127]: yeah, it's in openssl's error descriptions (which are different from boringssl ones).

File: test/extensions/filters/http/lua/lua_integration_test.cc [hunk 8]
```diff
@@ -477,6 +477,7 @@ name: lua
   cleanup();
 }
 
+// TODO (dmitri-d) re-enable when importPublicKey method is added
```
Comment (knrc) [inline test/extensions/filters/http/lua/lua_integration_test.cc:480]: Was this fixed?

File: test/extensions/transport_sockets/tls/ssl_socket_test.cc [hunk 9]
```diff
@@ -271,6 +270,8 @@ void testUtil(const TestUtilOptions& options) {
   ON_CALL(server_factory_context, api()).WillByDefault(ReturnRef(*server_api));
 
   // For private key method testing.
+  // TODO (dmitri-d) This is currently not supported under openssl
+/*
```
[the remark this answers is not in the document]
Comment (dmitri-d) [reply test/extensions/transport_sockets/tls/ssl_socket_test.cc:274]: No, this hasn't been cleaned up, pls. also see my comment here: https://github.com/Maistra/envoy/pull/13#discussion_r423299853

**Comment by dmitri-d** [comment]
* responded to feedback

File: bazel/repository_locations.bzl [hunk 10]
```diff
@@ -183,10 +184,10 @@ REPOSITORY_LOCATIONS = dict(
         urls = ["https://github.com/msgpack/msgpack-c/releases/download/cpp-3.2.1/msgpack-3.2.1.tar.gz"],
     ),
     com_github_google_jwt_verify = dict(
-        sha256 = "d422a6eadd4bcdd0f9b122cd843a4015f8b18aebea6e1deb004bd4d401a8ef92",
-        strip_prefix = "jwt_verify_lib-40e2cc938f4bcd059a97dc6c73f59ecfa5a71bac",
-        # 2020-02-11
-        urls = ["https://github.com/google/jwt_verify_lib/archive/40e2cc938f4bcd059a97dc6c73f59ecfa5a71bac.tar.gz"],
+        sha256 = "974826c9ba8060d9656e6c360718e3ee5eb1622ff59ed4138f8a3cd718ea7b13",
+        strip_prefix = "jwt_verify_lib-2c7ddcd871c3d902e089787e244ebe49723652cc",
+        # 2019-12-20
```
Comment (brian-avery) [inline bazel/repository_locations.bzl:189]: What's the date here? Commit is dated February 11
https://github.com/Maistra/jwt_verify_lib/commit/2c7ddcd871c3d902e089787e244ebe49723652cc

Comment (dmitri-d) [reply bazel/repository_locations.bzl:189]: It's an old one, I have a separate PR that updates jwt lib to a more recent version.

Comment (dmitri-d) [reply bazel/repository_locations.bzl:189]: The comment is from maistra-1.1 branch: https://github.com/Maistra/envoy/blob/maistra-1.1/bazel/repository_locations.bzl#L180

File: source/extensions/transport_sockets/tls/openssl_impl.cc [hunk 11]
```diff
@@ -0,0 +1,173 @@
...
+  if (SSL_select_next_proto(const_cast<unsigned char**>(out), outlen, alpn_data, alpn_data_size, in,
+                            inlen) != OPENSSL_NPN_NEGOTIATED) {
+    return SSL_TLSEXT_ERR_NOACK;
+  } else {
+    return SSL_TLSEXT_ERR_OK;
+  }
+}
+
+auto select_client_cert_cb = +[](SSL* ssl, X509** x509, EVP_PKEY** pkey) -> int {
+  std::cout << "!!!!!!!!!!!!!!!!!!!!!!!!!!!!! select_client_cert_cb \n";
```
Comment (brian-avery) [inline source/extensions/transport_sockets/tls/openssl_impl.cc:36]: is this intended?

Comment (dmitri-d) [reply source/extensions/transport_sockets/tls/openssl_impl.cc:36]: This is from the original openssl work. I removed it when I made changes for this release (next PR).

**Pull request merged**